// astro/dist/core/render/dev/index.js
import { PAGE_SCRIPT_ID } from "../../../vite-plugin-scripts/index.js";
import { enhanceViteSSRError } from "../../errors/dev/index.js";
import { AggregateError, CSSError, MarkdownError } from "../../errors/index.js";
import { isPage, resolveIdToUrl, viteID } from "../../util.js";
import { createRenderContext, loadRenderers, tryRenderPage } from "../index.js";
import { getStylesForURL } from "./css.js";
import { getComponentMetadata } from "./metadata.js";
import { getScriptsForURL } from "./scripts.js";
import { createDevelopmentEnvironment } from "./environment.js";
async function preload({
  env,
  filePath
}) {
  const renderers = await loadRenderers(env.settings, env.loader);
  env.renderers = renderers;
  try {
    const mod = await env.loader.import(viteID(filePath));
    return mod;
  } catch (error) {
    if (MarkdownError.is(error) || CSSError.is(error) || AggregateError.is(error)) {
      throw error;
    }
    throw enhanceViteSSRError({ error, filePath, loader: env.loader });
  }
}
async function getScriptsAndStyles({ env, filePath }) {
  const scripts = await getScriptsForURL(filePath, env.settings.config.root, env.loader);
  if (isPage(filePath, env.settings) && env.mode === "development") {
    scripts.add({
      props: { type: "module", src: "/@vite/client" },
      children: ""
    });
    scripts.add({
      props: {
        type: "module",
        src: await resolveIdToUrl(env.loader, "astro/runtime/client/hmr.js")
      },
      children: ""
    });
  }
  for (const script of env.settings.scripts) {
    if (script.stage === "head-inline") {
      scripts.add({
        props: {},
        children: script.content
      });
    } else if (script.stage === "page" && isPage(filePath, env.settings)) {
      scripts.add({
        props: { type: "module", src: `/@id/${PAGE_SCRIPT_ID}` },
        children: ""
      });
    }
  }
  const { urls: styleUrls, stylesMap } = await getStylesForURL(filePath, env.loader, env.mode);
  let links = /* @__PURE__ */ new Set();
  [...styleUrls].forEach((href) => {
    links.add({
      props: {
        rel: "stylesheet",
        href
      },
      children: ""
    });
  });
  let styles = /* @__PURE__ */ new Set();
  [...stylesMap].forEach(([url, content]) => {
    scripts.add({
      props: {
        type: "module",
        src: url
      },
      children: ""
    });
    styles.add({
      props: {
        type: "text/css",
        // Track the ID so we can match it to Vite's injected style later
        "data-astro-dev-id": viteID(new URL(`.${url}`, env.settings.config.root))
      },
      children: content
    });
  });
  const metadata = await getComponentMetadata(filePath, env.loader);
  return { scripts, styles, links, metadata };
}
async function renderPage(options) {
  var _a;
  const mod = options.preload;
  const { scripts, links, styles, metadata } = await getScriptsAndStyles({
    env: options.env,
    filePath: options.filePath
  });
  const { env } = options;
  const renderContext = await createRenderContext({
    request: options.request,
    pathname: options.pathname,
    scripts,
    links,
    styles,
    componentMetadata: metadata,
    route: options.route,
    mod,
    env
  });
  const onRequest = (_a = options.middleware) == null ? void 0 : _a.onRequest;
  return tryRenderPage(renderContext, env, mod, onRequest);
}
export {
  createDevelopmentEnvironment,
  preload,
  renderPage
};
